### PR TITLE
Inline functions (including multi-line and non-elemental functions)

### PR DESCRIPTION
**Pull request by MichaelSt98**
Extending and improving Inline utilities as discussed in #377

### CODE REVIEW

**Comment by MichaelSt98** [comment]
The whole thing got a bit out of hand again and turned out to be a rabbit hole ...

Currently (or rather with this PR), nested functions are handled via an iterative approach. We could try to solve this recursively instead. However, I don't think this would be straight-forward and questionable whether this is beneficial. Anyway, I am open for discussing that and other implementation details.

Moreover, it would be possible to check whether functions are single-line (or could be reduced to a single line) and inline those at an expression level.

**Review by mlange05 — APPROVED** [review]
Looks great, and is very, very useful! Many thanks, this one has been on the todo list for a while! 🙏 🙏 🙏 🙏 🙏 🙏  GTG :shipit:

File: loki/transformations/tests/test_inline.py [hunk 1]
```diff
@@ -367,6 +465,81 @@ def test_inline_member_routines(tmp_path, frontend):
     assert (b == [3., 3., 3.]).all()
 
 
+@pytest.mark.parametrize('frontend', available_frontends(skip={
+    OMNI: "OMNI can't parse the code"}))
```
Comment (reuterbal) [inline loki/transformations/tests/test_inline.py:469]: Why? I can't see anything that would usually upset OMNI, and a quick parse with `F_Front` was successful for me.

File: loki/transformations/tests/test_inline.py [hunk 2]
```diff
@@ -367,6 +465,81 @@ def test_inline_member_routines(tmp_path, frontend):
...
+def test_inline_member_functions(tmp_path, frontend):
+    """
+    Test inlining of member subroutines.
+    """
+    fcode = """
+subroutine member_functions(a, b, c)
```
Comment (reuterbal) [inline loki/transformations/tests/test_inline.py:475]:
```suggestion
subroutine member_functions(a, b, c)
  implicit none
```

**Review by reuterbal — CHANGES_REQUESTED** [review]
Very nicely done, this is a really fantastic addition and nicely executed!

A bunch of nitpick comments but no showstoppers.

File: loki/transformations/tests/test_inline.py (hunk 1 above)
Comment (MichaelSt98) [reply loki/transformations/tests/test_inline.py:469]: If you execute

```py
from loki import Subroutine, fgen, OMNI, FP
fcode = """
subroutine member_functions
    implicit none
    integer :: i
    real(kind=8) :: a(3)
    contains
    function add_to_a(b, n)
      integer, intent(in) :: n
      real(kind=8), intent(in) :: b(n)
      real(kind=8) :: add_to_a(n)

      do i = 1, n
        add_to_a(i) = a(i) + b(i)
      end do
    end function
end subroutine member_functions
    """
routine = Subroutine.from_source(fcode, frontend=OMNI)
print(fgen(routine))
```

```fortran
SUBROUTINE member_functions ()
  IMPLICIT NONE
  INTEGER :: i
  REAL(KIND=8) :: a(3)
  CONTAINS
  FUNCTION add_to_a (b, n)
    IMPLICIT NONE
    INTEGER, INTENT(IN) :: n
    REAL(KIND=8), INTENT(IN) :: b(n)
    ! HERE: MISSING DIMENSION !!!
    REAL(KIND=8) :: add_to_a
    DO i=1,n
      add_to_a(i) = a(i) + b(i)
    END DO
  END FUNCTION add_to_a
END SUBROUTINE member_functions
```

Same thing for `real(kind=8), dimension(n) :: add_to_a`. This doesn't happen for `FP` and `OFP`.

Comment (reuterbal) [reply loki/transformations/tests/test_inline.py]: #391 should fix the problem

**Review by reuterbal — APPROVED** [review]
I have rebased and removed the pylintrc commit. Ready from my side now, many thanks again!

**Pull request merged**